### PR TITLE
[IOTDB-2075] Accelerate the process of insertTablets by using thread pool

### PR DESCRIPTION
**Pull request by jun0315**
See https://issues.apache.org/jira/browse/IOTDB-2075.

Performance report: TBD.

### CODE REVIEW

**Comment by jun0315** [comment]
test result:

tablet count	use serial	use thread pool
100	169	166
1000	378	308
5000	1044	918
10000	1375	1497
50000	7648	7172

Because the speed of insertTable is very fast, about 0.15ms, there is no acceleration effect. In order to verify this conjecture,
insertTable sleep 10 ms

tablet count	use serial	use thread pool
100	1206	216
1000	12310	1167
5000	59562	5177


test device:
MacBook Pro (16-inch, 2019)
2.6 GHz 6-core processor Core i7
16 GB 2667 MHz DDR4

**Comment by jun0315** [comment]
# Test Result



|                             | Serialization | 2 corePoolSize | 4 corePoolSize | 6 corePoolSize | 8 corePoolSize | 10 corePoolSize | 12 corePoolSize |
| --------------------------- | ------------- | -------------- | -------------- | -------------- | -------------- | --------------- | --------------- |
| 1000 Plans 48 rows  16 columns | 1397          | 1292           | 1160           | 1040           | 1085           | 1089            | 1102            |

We can find that the efficiency is the highest when we take Runtime.getRuntime().availableProcessors()/2.

Next, we default to 1000 Plans 48 rows and change the different columns.


|                    | Serialization | 6 corePoolSize |
| ------------------ | ------------- | -------------- |
| 4 columns      | 901           | 1023           |
| 8 columns      | 958           | 1068           |
| 12 columns     | 1196          | 1259           |
| 14 columns     | 1278          | 1239           |
| **15 columns** | **1215**      | **1001**       |
| **16 columns** | **1364**      | **1014**       |


We can find that when there are 15 columns and 16 columns, the thread pool speed will exceed the serial speed, so we set the threshold to 15.

To sum up, we set the number of core threads in the thread pool to min(the number of different sg, availableProcessors()/2), and need columns >= 15

## Test Device

MacBook Pro (16-inch, 2019)

2.6 GHz 6-core processor Core i7

16 GB 2667 MHz DDR4

**Comment by jun0315** [comment]
Since the default maximum row is 10000, we redo the following experiment with row = 10000
# Test Result

|                                | Serialization | 2 corePoolSize | 4 corePoolSize | 6 corePoolSize | 8 corePoolSize | 10 corePoolSize | 12 corePoolSize |
| ------------------------------ | ------------- | -------------- | -------------- | -------------- | -------------- | --------------- | --------------- |
| 1000 Plans 10000 row 16 column | 1952          | 1906           | 1742           | 1608           | 1736           | 1670            | 1641            |

We can also find Runtime.getRuntime().availableProcessors() = 12 on known test machine

We can find that the efficiency is the highest when we take Runtime.getRuntime().availableProcessors()/2.


|                              | Serialization | 6 corePoolSize |
| ---------------------------- | ------------- | -------------- |
| 1000Plans 10000row 8 column | 710         | 809         |
| 1000Plans 10000row 10 column              | 1018        | 960       |
| 1000Plans 10000row 12 column              | 1912       | 1717        |

We can find that when there are 10 columns , the thread pool speed will exceed the serial speed, so we set the threshold to 10.

Therefore, we set the number of core threads in the thread pool to min(the number of different sg, availableProcessors()/2), and need columns >= 10

It should be noted that in the latest test, we found that if the number of sg is large and exceeds twice the recommended number of CPU threads, it **may** lead to failure to allocate out of heap memory and NPE. Therefore, we will also turn off multithreading in this case.

## Test Device
MacBook Pro (16-inch, 2019)

2.6 GHz 6-core processor Core i7

16 GB 2667 MHz DDR4

File: server/src/assembly/resources/conf/iotdb-engine.properties [hunk 1]
```diff
@@ -884,6 +884,15 @@ timestamp_precision=ms
...
 # select_into_insert_tablet_plan_row_limit=10000
 
+
+####################
+### Insert-Tablets Configuration
+####################
+
+# When the insert plan rowcount reaches a certain quantity and exceeds a certain ratio, turn on Multithreading acceleration
+# Datatype: int
+# insert_multi_tablet_enable_thread_pool_row_count_threshold=10
```
Comment (SteveYurongSu) [inline server/src/assembly/resources/conf/iotdb-engine.properties:894]:
```suggestion
# insert_multi_tablet_enable_multithreading_column_threshold=10
```

File: server/src/main/java/org/apache/iotdb/db/concurrent/ThreadName.java [hunk 2]
```diff
@@ -44,6 +44,7 @@
   LOAD_TSFILE("Load-TsFile"),
   TIME_COST_STATISTIC("TIME_COST_STATISTIC"),
   QUERY_SERVICE("Query"),
+  INSERT_SERVICE("InsertTablets"),
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/concurrent/ThreadName.java:47]:
```suggestion
  INSERTION_SERVICE("MultithreadingInsertionPool"),
```

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertMultiTabletPlan.java [hunk 3]
```diff
@@ -384,4 +388,52 @@ public void unsetIsExecuted(int i) {
       results.remove(i);
     }
   }
+
+  public int getInsertPlanSGSize() {
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertMultiTabletPlan.java:392]:
```suggestion
  public int getDifferentStorageGroupsCount() {
```

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertMultiTabletPlan.java [hunk 4]
```diff
@@ -96,6 +96,10 @@
 
   boolean[] isExecuted;
 
+  Boolean needMulti;
+
+  Integer insertPlanSGSize;
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertMultiTabletPlan.java:101]:
```suggestion
  Boolean isEnableMultithreading;

  Integer differentStorageGroupsCount;
```

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertMultiTabletPlan.java [hunk 5]
```diff
@@ -384,4 +388,52 @@ public void unsetIsExecuted(int i) {
...
+        } catch (MetadataException ignored) {
+        }
+      }
+      insertPlanSGSize = insertPlanSGSet.size();
+    }
+    return insertPlanSGSize;
+  }
+
+  public boolean needMultiThread() {
+    // After testing, we can find that when there are 10 columns , the thread pool speed will exceed
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertMultiTabletPlan.java:410]:
```suggestion
    if (IoTDB.isClusterMode()) {
        isEnableMultithreading = false;
        return isEnableMultithreading;
    }
    // After testing, we can find that when there are 10 columns , the thread pool speed will exceed
```

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 6]
```diff
@@ -1524,8 +1540,56 @@ public void insertTablet(InsertMultiTabletPlan insertMultiTabletPlan)
...
+    }
+  }
+
+  private void updateInsertTabletsPool(int sgSize) {
+    int updateCoreSize = Math.min(sgSize, Runtime.getRuntime().availableProcessors() / 2);
+    if (insertionPool == null || insertionPool.isTerminated()) {
+      insertionPool =
+          (ThreadPoolExecutor)
+              IoTDBThreadPoolFactory.newFixedThreadPool(
+                  Runtime.getRuntime().availableProcessors(), ThreadName.INSERT_SERVICE.getName());
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:1583]:
```suggestion
                  updateCoreSize, ThreadName.INSERT_SERVICE.getName());
```

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 7]
```diff
@@ -1524,8 +1540,56 @@ public void insertTablet(InsertMultiTabletPlan insertMultiTabletPlan)
...
+  private void insertTabletParallel(InsertMultiTabletPlan insertMultiTabletPlan)
+      throws BatchProcessException {
+    updateInsertTabletsPool(insertMultiTabletPlan.getDifferentStorageGroupsCount());
+
+    List<InsertTabletPlan> planList = insertMultiTabletPlan.getInsertTabletPlanList();
+    List<Future<?>> futureList = new ArrayList<>();
+
+    Map<Integer, TSStatus> results = insertMultiTabletPlan.getResults();
+    for (InsertTabletPlan plan : planList) {
+      Future<?> f =
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:1552]:
```suggestion
      Future<Exception> f =
```

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 8]
```diff
@@ -1524,8 +1540,56 @@ public void insertTablet(InsertMultiTabletPlan insertMultiTabletPlan)
...
+          insertionPool.submit(
+              () -> {
+                insertTablet(plan);
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:1555]:
```suggestion
                try {
                  insertTablet(plan);
                } catch (Exception e) {
                  return e;
                }
```

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 9]
```diff
@@ -1524,8 +1540,56 @@ public void insertTablet(InsertMultiTabletPlan insertMultiTabletPlan)
...
+                return null;
+              });
+      futureList.add(f);
+    }
+    for (int i = 0; i < futureList.size(); i++) {
+      try {
+        futureList.get(i).get();
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:1562]:
```suggestion
        futureList.get(i).get();
```
Handle the exception here :D

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 10]
```diff
@@ -1524,8 +1540,56 @@ public void insertTablet(InsertMultiTabletPlan insertMultiTabletPlan)
...
+      } catch (Exception e) {
+        if (e.getCause() instanceof QueryProcessException) {
+          QueryProcessException qe = (QueryProcessException) e.getCause();
+          results.put(i, RpcUtils.getStatus(qe.getErrorCode(), qe.getMessage()));
+        } else {
+          results.put(i, RpcUtils.getStatus(TIME_OUT, e.getMessage()));
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:1568]: The exception may not only be a TIME_OUT exception.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertMultiTabletPlan.java [hunk 11]
```diff
@@ -384,4 +388,59 @@ public void unsetIsExecuted(int i) {
...
+  public int getDifferentStorageGroupsCount() {
+    if (differentStorageGroupsCount == null) {
+      Set<String> insertPlanSGSet = new HashSet<>();
+      for (InsertTabletPlan insertTabletPlan : insertTabletPlanList) {
+        IStorageGroupMNode storageGroupMNode = null;
+        try {
+          storageGroupMNode =
+              IoTDB.metaManager.getStorageGroupNodeByPath(insertTabletPlan.getDeviceId());
+          insertPlanSGSet.add(storageGroupMNode.getFullPath());
+        } catch (MetadataException ignored) {
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertMultiTabletPlan.java:401]: This can not be ignored. Throwing a RuntimeException here is OK.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertMultiTabletPlan.java [hunk 12]
```diff
@@ -384,4 +388,59 @@ public void unsetIsExecuted(int i) {
...
     }
   }
+
+  public int getDifferentStorageGroupsCount() {
+    if (differentStorageGroupsCount == null) {
+      Set<String> insertPlanSGSet = new HashSet<>();
+      for (InsertTabletPlan insertTabletPlan : insertTabletPlanList) {
+        IStorageGroupMNode storageGroupMNode = null;
+        try {
+          storageGroupMNode =
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertMultiTabletPlan.java:398]:
```suggestion
        try {
          IStorageGroupMNode storageGroupMNode =
```

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertMultiTabletPlan.java [hunk 13]
```diff
@@ -384,4 +388,59 @@ public void unsetIsExecuted(int i) {
...
+        }
+      }
+      differentStorageGroupsCount = insertPlanSGSet.size();
+    }
+    return differentStorageGroupsCount;
+  }
+
+  public boolean isEnableMultiThreading() {
+    // After testing, we can find that when there are 10 columns , the thread pool speed will exceed
+    // the serial speed, so we set the threshold to 10.
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertMultiTabletPlan.java:411]:
```suggestion
```
Explain why need the InsertMultiTabletEnableMultithreadingColumnThreshold configuration parameter here.

**Review by SteveYurongSu — CHANGES_REQUESTED** [review]
Much better this time :D

**Review by SteveYurongSu — APPROVED** [review]
LGTM

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 14]
```diff
@@ -1524,8 +1540,72 @@ public void insertTablet(InsertMultiTabletPlan insertMultiTabletPlan)
...
+          insertionPool.submit(
+              () -> {
+                insertTablet(plan);
+                return null;
+              });
+      futureList.add(f);
+    }
+    for (int i = 0; i < futureList.size(); i++) {
+      try {
+        futureList.get(i).get();
```
Comment (dake0805) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:1571]: Future.get() should set timeout param , or else it may cause blocking.

Comment (jun0315) [reply server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:1571]: Thank you for your reiew, because this is a database system. If there is a blockage, there is usually a bug in the system. At this time, we hope to throw a problem to check the bug. Maybe we don't need to set timeout param, which may be different from the business system in the company.

**Review by SteveYurongSu — APPROVED** [review]
LGTM
